### PR TITLE
PISTON-822: Crossbar module for Tray.io

### PR DESCRIPTION
**Pull request by DustinBrett**
This module is a go-between for creating 3rd party integrations via Tray.io. This module handles creating tray user id's based on account id and also creating authorization codes which are used to edit what tray calls solution instances. Voxter is working on a front-end Monster UI app that will use these endpoints to make Tray.io solution instances for accounts. This will allow things such as "When a call comes in message a channel in Slack". To use this module a Tray.io account is required.

### CODE REVIEW

**Comment by DustinBrett** [comment]
Fixed the build error involving mkdocs.yml. The current build error doesn't seem to be related to this PR. (https://circleci.com/gh/2600hz/kazoo/26188)

**Comment by DustinBrett** [comment]
> @DustinBrett this looks really cool, but we should go over the `code` for handling access/authorization. will you be @kazoocon ?

Yes I will be there and I'd be happy to discuss it.

**Comment by DustinBrett** [comment]
> > @DustinBrett this looks really cool, but we should go over the `code` for handling access/authorization. will you be @kazoocon ?
> 
> Yes I will be there and I'd be happy to discuss it.

Hi @lazedo,

KazooCon was a lot of fun but unfortunately I never found the right time to discuss your concerns about my PR. Could we perhaps discuss them now in these comments? I believe your concerns were about "handling access/authorization". Thanks.

**Comment by DustinBrett** [comment]
I see my build failed because of creation of unstaged files:
- core/kazoo_documents/src/kzd_tray_create.erl.src
- core/kazoo_documents/src/kzd_tray_toggle.erl.src

Am I using schemas incorrectly or should I create these files? Thanks.

**Comment by DustinBrett** [comment]
Since originally creating this PR I have done several changes based on feedback locally.
- Added execute bindings where applicable instead of doing it all after validation
- Switched to endpoint validation via schemas (This needs to be discussed because the build pipeline requires a lot of additional changes for this to work)
- Added "transcribed" boolean to message list requests so that we can know if a voicemail has a transcription associated

**Comment by jamesaimonetti** [comment]
> I see my build failed because of creation of unstaged files (https://circleci.com/gh/2600hz/kazoo/26420):
> 
>     * core/kazoo_documents/src/kzd_tray_create.erl.src
> 
>     * core/kazoo_documents/src/kzd_tray_toggle.erl.src
> 
> 
> Am I using schemas incorrectly or should I create these files? Thanks.

I will add docs here but:

We parse the schemas and create these .src accessor modules. When you have a new .src you need to copy it to a .erl and commit it. When changes happen to the schema, the .src will get updated and will need to be merged into the .erl. This let's us add functionality to the .erl without needing to automate the merging of schema changes.

File: applications/crossbar/src/crossbar.hrl [hunk 1]
```diff
@@ -121,6 +121,7 @@
                          ,'cb_token_auth'
                          ,'cb_token_restrictions'
                          ,'cb_transactions'
+                         ,'cb_tray'
```
Comment (jamesaimonetti) [inline applications/crossbar/src/crossbar.hrl]: I don't think this should be added to the default list. It should be explicitly enabled by sysadmins, including new installs.

**Comment by DustinBrett** [comment]
> > I see my build failed because of creation of unstaged files (https://circleci.com/gh/2600hz/kazoo/26420):
> > ```
> > * core/kazoo_documents/src/kzd_tray_create.erl.src
> > 
> > * core/kazoo_documents/src/kzd_tray_toggle.erl.src
> > ```
> > 
> > 
> > Am I using schemas incorrectly or should I create these files? Thanks.
> 
> I will add docs here but:
> 
> We parse the schemas and create these .src accessor modules. When you have a new .src you need to copy it to a .erl and commit it. When changes happen to the schema, the .src will get updated and will need to be merged into the .erl. This let's us add functionality to the .erl without needing to automate the merging of schema changes.

Ok thanks, that's good to know. So I can create these erl/src files, but before I do that, am I correct to be using the schema in this way? There is not document that matches this schema but I created it so that I could validate my endpoints which require certain data. Thanks.

File: applications/crossbar/src/crossbar.hrl (hunk 1 above)
Comment (DustinBrett) [reply applications/crossbar/src/crossbar.hrl]: Ok that makes sense, I will remove this.

**Comment by DustinBrett** [comment]
Sorry for this branch being a bit messy. Mistakes were made with a git command. Cleaned up but left unsquashed for now until I have all feedback. Once I have all feedback I will redo this all and clean everything up.

**Comment by DustinBrett** [comment]
Finished updates and changes to code. Still waiting to hear from @lazedo about his `handling access/authorization` concerns. Otherwise it's ready.

**Comment by jamesaimonetti** [comment]
@DustinBrett can you rebase to latest master to clear up the CI passes?

**Comment by DustinBrett** [comment]
> @DustinBrett can you rebase to latest master to clear up the CI passes?

Okay sure np. I will do that as soon as I have a moment.

**Comment by DustinBrett** [comment]
> > @DustinBrett can you rebase to latest master to clear up the CI passes?
> 
> Okay sure np. I will do that as soon as I have a moment.

I've completed the rebase.

**Comment by DustinBrett** [comment]
I will look into using `kz_auth_client` and on why the Travis CI build kept failing. Will open a new PR in the future as this one seems to have lost traction.

**Pull request closed without merging**